spec/models/concerns/api_authenticable_spec.rb

```diff
@@ -12,4 +12,23 @@
       expect(user.authentication_token.length).to eq(ApiAuthenticable::AUTHENTICATION_TOKEN_LENGTH)
     end
   end
+
+  describe "valid_token?" do
+    it "Securely compares the tokens (prefixed with 'Token ')" do
+      user = build(:user, authentication_token: "SecurelyRandomToken")
+
+      allow(ActiveSupport::SecurityUtils).to receive(:secure_compare)
+        .and_return(:secure_compare_boolean_result)
+
+      expect(user.valid_token?("Token SecurelyRandomToken")).to eq :secure_compare_boolean_result
+      expect(ActiveSupport::SecurityUtils).to have_received(:secure_compare)
+        .with("Token SecurelyRandomToken", "Token SecurelyRandomToken")
```

Comment:
testing that it returns the symbol feels weird, we don't care about that in this spec, what we care about is the token and how it compares:
```suggestion
      user = build(:user, authentication_token: "SecurelyRandomToken")
      token = "SecurelyRandomToken"

      allow(ActiveSupport::SecurityUtils).to receive(:secure_compare)

      user.valid_token?("Token #{token}")
      expect(ActiveSupport::SecurityUtils).to have_received(:secure_compare)
        .with("Token SecurelyRandomToken", "Token SecurelyRandomToken")
```

Reply:
But we do need to care that the result of the `secure_compare` is what is being returned, so I think it is necessary.

locally, with this change applied, try adding a `true` as the return value for `valid_token?` and see if any tests fail.

Reply:
I use `:secure_compare_boolean_result` as a stub return value here because we need to check that that value is being returned and we don't care in this test whether it is true or false.

We could write two tests, one for token matching and one for token not matching, but if we don't want to call the real method we need to fake the return value twice as well, and what's the point of doing it twice if you only need to do it once? 🤷

Reply:
fair, actually testing that `ActiveSupport::SecureUtils` works feels silly 🤔